components/director/internal/tenantfetcher/service.go

```diff
@@ -88,84 +154,103 @@ func (s Service) SyncTenants() error {
 
-	currentTenants, err := s.tenantStorageService.List(ctx)
-	if err != nil {
-		return errors.Wrap(err, "while listing tenants")
-	}
+	existingTenantsSup := s.newCurrentTenantsSupplier()
 
-	currentTenantsMap := make(map[string]bool)
-	for _, ct := range currentTenants {
-		currentTenantsMap[ct.ExternalTenant] = true
+	actions := make([]*syncTenantsAction, 0)
+	if len(tenantsToCreate) == 0 && len(tenantsToDelete) == 0 {
+		actions = append(actions, &syncTenantsAction{"while moving runtimes by labels", func() error { return s.moveRuntimesByLabel(ctx, runtimesToMove) }})
+	} else {
+		actions = append(actions,
+			&syncTenantsAction{"while storing tenants", func() error { return s.createTenants(ctx, existingTenantsSup, tenantsToCreate) }},
+			&syncTenantsAction{"while moving runtimes by label", func() error { return s.moveRuntimesByLabel(ctx, runtimesToMove) }},
+			&syncTenantsAction{"while deleting tenants", func() error { return s.deleteTenants(ctx, existingTenantsSup, tenantsToDelete) }})
 	}
 
-	for i := len(tenantsToCreate) - 1; i >= 0; i-- {
-		if currentTenantsMap[tenantsToCreate[i].ExternalTenant] {
-			tenantsToCreate = append(tenantsToCreate[:i], tenantsToCreate[i+1:]...)
+	for _, action := range actions {
+		if err = action.Execute(); err != nil {
+			return errors.Wrap(err, "while processing events")
 		}
 	}
 
-	tenantsToDelete = make([]model.BusinessTenantMappingInput, 0)
-	for _, toDelete := range deleteTenantsMap {
-		if currentTenantsMap[toDelete.ExternalTenant] {
-			tenantsToDelete = append(tenantsToDelete, toDelete)
-		}
+	if err = tx.Commit(); err != nil {
+		return err
 	}
+	if err = s.kubeClient.UpdateTenantFetcherConfigMapData(ctx, convertTimeToUnixNanoString(startTime)); err != nil {
+		return err
+	}
+
+	return nil
+}
 
-	err = s.tenantStorageService.CreateManyIfNotExists(ctx, tenantsToCreate)
+func (s Service) createTenants(ctx context.Context, ts *currentTenantsSupplier, eventsTenants []model.BusinessTenantMappingInput) error {
+	currTenants, err := ts.Get(ctx)
 	if err != nil {
-		return errors.Wrap(err, "while storing new tenants")
+		return err
 	}
-	err = s.tenantStorageService.DeleteMany(ctx, tenantsToDelete)
-	if err != nil {
-		return errors.Wrap(err, "while removing tenants")
+
+	tenantsToCreate := make([]model.BusinessTenantMappingInput, 0)
+	for i := len(eventsTenants) - 1; i >= 0; i-- {
+		if currTenants[eventsTenants[i].ExternalTenant] {
+			continue
+		}
+		tenantsToCreate = append(tenantsToCreate, eventsTenants[i])
 	}
 
-	err = tx.Commit()
-	if err != nil {
-		return err
+	if err = s.tenantStorageService.CreateManyIfNotExists(ctx, tenantsToCreate); err != nil {
+		return errors.Wrap(err, "while storing new tenants")
 	}
 
-	err = s.kubeClient.UpdateTenantFetcherConfigMapData(ctx, convertTimeToUnixNanoString(startTime))
+	return nil
+}
+
+func (s Service) deleteTenants(ctx context.Context, ts *currentTenantsSupplier, eventsTenants []model.BusinessTenantMappingInput) error {
```

Comment:
Could you place the `moveRuntimesByLabel` method below this `deleteTenants` so as to sort them in a logical way as this file is already pretty overcrowded and there are numerous functions/methods everywhere?